[WFLY-10879] Deployment is not failing as expected and according to the specification if a @Singleton @Startup @PostConstruct initialization failed

spyrkob opened the pull request:
JIRA: https://issues.jboss.org/browse/WFLY-10879

@dmlloyd could you please take a look at this? The first commit uses StartupCountInterceptor to block access to EJBs unless all @Startup beans are successfully initiated. The second uses similar approach for web components

spyrkob committed:
[WFLY-10879] Blocking access to all EJB components if one of Startup
beans failed

spyrkob committed:
[WFLY-10879] Blocking undertow requests until all EJB startup components
initialize succesfully

spyrkob commented:
@dmlloyd did you have a chance to look into this?

dmlloyd commented on ee/src/main/java/org/jboss/as/ee/component/deployers/StartupCountdown.java:
```diff
@@ -47,6 +47,22 @@ public void await() throws InterruptedException {
+    if (isPrivileged()) return true;
+    if (count != 0) {
+      synchronized (this) {
+        long start = System.currentTimeMillis();
```
These constructs should always be using System.nanoTime(), not currentTimeMillis().

spyrkob replied:
Thanks, updated to use nanoTime

dmlloyd reviewed (changes requested):
Otherwise this seems OK.

dmlloyd commented on undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        boolean componentsStarted = countdown.tryAwait(COMPONENT_STARTUP_TIMEOUT, TimeUnit.MILLISECONDS);
```
Undertow handlers should never block.  This can cause the server to hang.  Instead the requests have to be queued somehow and replayed once the server finishes starting.  Also I'm not really sure that a 404 is the right response if the server takes too long to start.

spyrkob replied:
Yeah I wasn't sure about that part. I'll take a look at it tomorrow and try to come up with a better solution, thanks.

dmlloyd reviewed (dismissed):
Sorry, there's one more problem.  I didn't notice it because it used to be that tests always ended up at the bottom of the PR, but I guess they don't anymore.

spyrkob committed:
[WFLY-10879] Queue requests in Undertow handler until components are
either started or fail startup

dmlloyd commented:
This is now too complex to review in a short amount of time; could you please summarize the operation of the code itself?

spyrkob commented:
1. `StartupCountdown` and `StartupCountdownInterceptor` is changed to invoke either:
   1. onSuccess callback - when all `@Startup` components finished initialization without exceptions, or
   2. onFailure callback - when any of components fails to initialize
2. `ComponentStartupCountdownHandler` is added to Undertow handlers stack and uses `SuspendRequestQueue` to suspend requests based on `StartupCountdown` callbacks: 
   1. if neither `onSuccess` or `onFailure` were called (ie. components are still starting) - will queue up requests in memory,
   2. if `StartupCountdown` invoked `onFailure` callback - will resume all queued up requests, but instead of processing them as normal, they will be failed. Same will apply to any further request for this deployment,
    3. if `StartupCountdown` invoked `onSuccess` callback - will resume all queued up requests dispatching them to to their remaining handlers. Any further request for this deployment will skip the queue and will be passed to next handler in chain
3. `ModuleDeployment` will prevent the EJB components from being started if `StartupCallback` has not emitted onSuccess callback.

bstansberry reviewed (commented):
The ComponentStartupCountdownHandler part seems redundant. We start up gracefully by default, and if people want to redeploy an app without reloading/restarting the server they can and should suspend it first. IOW having in-flight requests going while an app is starting is avoidable and not recommended, and we shouldn't be adding complex code to handle people doing things we don't recommend.

spyrkob committed:
[WFLY-10879] changing ComponentStartupCountdownHandler to rejects requests when
waiting for components to startup

spyrkob commented:
sorry for delay on this.
I rebased and removed request queuing. Since there was no need any more for failed/succeeded callbacks in StartupCountdown I restored it to original form.

If this looks OK I'll squash the commits.

darranl commented:
Retest this please

darranl commented:
@tadamski / @chengfang Could you please add a review to this PR - it is presently the oldest PR in the queue being over a year old so we either need to take this to a conclusion or close the PR.

tadamski reviewed (approved):
This fix seems good to me.

spyrkob commented:
Rebased the PR. Hopefully that resolves the deps-changed label - there are no dependency changes in the PR

darranl commented:
Just kicked off CI again, we have approvals so if we get a clean run we should get this in.

spyrkob commented:
I don't know if those test failures are related - I wasn't able to reproduce them in my environment.
Should I rebase the PR to get rid of the deps-changed label?

spyrkob commented:
retest this please

spyrkob commented:
@darranl can I do anything else to get this merged?

spyrkob committed:
[WFLY-13515] Fix problem with init-in-order

spyrkob commented:
There's one more case that the PR doesn't cover - when the EAR specifies initialize-in-order in application.xml.

Added a commit fixing this problem. @chengfang, @dmlloyd, @tadamski could you please review again? 
@chengfang with the change this should address WFLY-13515 as well.

chengfang commented:
From July 17, 2020 test runs, Linux JDK11 and Linux JDK8 both failed with 12 test failures in `mixed-domain/src/test/java/org/jboss/as/test/integration/domain/mixed/eap700/MixedDomainDeployment700TestCase`, which seems related to the code change here. Windows JDK11 failed with execution timeout (no specific test failures shown), and I'll try to restart Windows JDK11 job.

spyrkob commented:
@chengfang I can't reproduce this failures locally - do you know if this is an env issue?

darranl commented:
The mixed domain failures are environmental, they are going to be temporarily disabled under this PR https://github.com/wildfly/wildfly/pull/13418

darranl commented:
Re the merge, Jeff is tagging WildFly Core today but will try and have another pass through the PR queue either the end of this week or early next and get approved PRs merged.

chengfang commented:
Looks good.

spyrkob commented:
@bstansberry, @dmlloyd I think your comments should be addressed now. Could you approve this if it looks good? There's a customer case related to this, so it'd be great to get this merged :)

spyrkob commented on ejb3/src/main/java/org/jboss/as/ejb3/component/singleton/StartupCountDownInterceptor.java:
```diff
@@ -22,10 +22,14 @@
+      return proceed;
+    } catch (Exception e) {
+      // do not decrease countdown
+      throw e;
```
removed, thanks

spyrkob commented:
@dmlloyd I created PR in Core to add a new constant: https://github.com/wildfly/wildfly-core/pull/4300. The tests will be failing until it is merged though

bstansberry commented on undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java:
```diff
@@ -0,0 +1,56 @@
+    private final HttpHandler wrappedHandler;
+
+    private final AtomicBoolean started = new AtomicBoolean(false);
+    private volatile HttpHandler startupFailedHandler = ResponseCodeHandler.HANDLE_500;
```
This should be final.

spyrkob replied:
fixed, thanks

spyrkob commented:
@bstansberry core upgrade with wildfly/wildfly-core#4300 have been applied, I think the label can be removed

bstansberry commented:
Thanks, @spyrkob ! You have great patience. ;)

jonmbake commented:
This changed caused a subtle bug for us. I wanted to post here in case anyone else runs into the same issue.

We had a EAR structured with module loading order (defined by application.xml) like:

1. ejb1.jar
  - Contains `StartupBean1`
2. web.war
  - Contains a `@WebListener` whose `#contextInitialized` calls into `StartupBean1`
3. ejb2.jar
  - Contains `StartupBean2`

When deploying the ear, it will indefinitely wait for `StartupBean2` to load on the web listener call to `StartupBean1`. It's easily fixable by changing the application.xml to load ejb2.jar before web.war, but it can be challenging figuring out the issue.
